########################################################################################################
## K-means clustering of the IPK genebank barley collection data, region of interest on chromosome 5H ##
########################################################################################################

# Note: To analyze the IPK genebank data, we utilized WGS data generated for all accessions. the original dataset was published in Milner et al., 2019. Here we used an updated dataset with all variants mapped to Morex V3 (Mascher et al., 2021) provided by Martin Mascher, IPK Gatersleben. 

source("https://raw.githubusercontent.com/rwonneberger/R_functions/main/Ronja_functions.R")
library(adegenet)
library(factoextra)
library(ggforce)

colors=c('#e6194B', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#42d4f4', '#f032e6', '#bfef45', '#fabed4', '#469990', '#dcbeff', '#9A6324', '#fffac8', '#800000', '#aaffc3', '#808000', '#ffd8b1', '#000075' )


# The 5H region of interest is located at ~70 - 320 Mbp in Barke. We extracted 70000000-70005000 and 320000000-320005000 from Barke and blasted the sequences against Morex V3 (Mascher et al., 2021).
# The start is ca 66.2 Mbp in Morex, we selected 67 Mbp as a more conservative approach. The end is ca 312.2 Mbp in Morex, we selected 312 Mbp as a more conservative approach.
# We only considered accessions that were "domesticated" according to their passport information.
# We filtered the dataset to remove indels, and only kept SNPs with a minor allele count of > 1 and missing data < 20%.
# The final dataset contained 6536 SNPs and 19838 individuals.
# This script requires the described dataset as input, which will need to be generated by the Filtering_genebank_SNP_data.sh script. In addition, the user will need the passport data of the gene bank collection.


#  To identify the best number of subpopulations in the population, run the code commented out here:

#########################################################################################

# All domesticated --------------------------------------------------------
# 
# hmp=read.table("Data/IPK_domesticated_5H_20missing_MAC.hmp.txt",stringsAsFactors=F, na.strings = "N", header = T, sep="\t", comment.char="")
# 
# metadata <- fread("Data/IPK_domesticated.txt")
# 
# metadata%<>%filter(accession != "BRIDGE_HOR_7066", accession != "BRIDGE_HOR_21050", accession != "BRIDGE_HOR_16618") # These will be outliers in the PCA so should be removed to facilitate visualization
# 
# SamplesKeep <- names(hmp)[-c(1:11)]
# metadata_reduced <- metadata[metadata$accession%in%SamplesKeep,]
# 
# 
# metadata_reduced<-as.data.frame(metadata_reduced)
# #metadata_reduced[is.na(metadata_reduced$Haplotype),15:20]="NA"
# 
# names(hmp)[1]<-"accession"
# loc<-transpose_df_Col1(hmp[,-c(2:11)], "accession")
# 
# loc<-left_join(metadata_reduced, loc)
# 
# locus<-loc[,-c(1:14)]
# ind <- names(hmp)[-c(1:11)]
# 
# 
# table(names(hmp)[-c(1:11)]  == loc$accession)
# 
# 
# 
# genind_obj <- df2genind(locus, ploidy = 2, ind.names = ind,  sep = "")
# genind_obj
# 
# 
# 
# # Check the memberships with different K
# 
# grp_200_2<-find.clusters(genind_obj, max.n.clust=50,  n.pca=200, n.clust=2) # 200 PCs, 2 K
# 
# grpout<-as.data.frame(grp_200_2$grp)
# grpout$accession<-rownames(grpout)
# grpout<-left_join(grpout, metadata_reduced, by="accession")
# fwrite(as.data.frame(grpout), "Results/IPK_All_domesticated_grp_200_2.txt", sep="\t", row.names=T)
# 
# grp_200_3<-find.clusters(genind_obj, max.n.clust=50,  n.pca=200, n.clust=3) # 200 PCs, 3 K
# 
# grpout<-as.data.frame(grp_200_3$grp)
# grpout$accession<-rownames(grpout)
# grpout<-left_join(grpout, metadata_reduced, by="accession")
# fwrite(as.data.frame(grpout), "Results/IPK_All_domesticated_grp_200_3.txt", sep="\t", row.names=T)
# 
# grp_200_4<-find.clusters(genind_obj, max.n.clust=50,  n.pca=200, n.clust=4) # 200 PCs, 4 K
# 
# grpout<-as.data.frame(grp_200_4$grp)
# grpout$accession<-rownames(grpout)
# grpout<-left_join(grpout, metadata_reduced, by="accession")
# fwrite(as.data.frame(grpout), "Results/IPK_All_domesticated_grp_200_4.txt", sep="\t", row.names=T)
# 
# grp_200_5<-find.clusters(genind_obj, max.n.clust=50,  n.pca=200, n.clust=5) # 200 PCs, 5 K
# 
# grpout<-as.data.frame(grp_200_5$grp)
# grpout$accession<-rownames(grpout)
# grpout<-left_join(grpout, metadata_reduced, by="accession")
# fwrite(as.data.frame(grpout), "Results/IPK_All_domesticated_grp_200_5.txt", sep="\t", row.names=T)
# 
# 
# grp_200_6<-find.clusters(genind_obj, max.n.clust=50,  n.pca=200, n.clust=6) # 200 PCs, 6 K
# 
# grpout<-as.data.frame(grp_200_6$grp)
# grpout$accession<-rownames(grpout)
# grpout<-left_join(grpout, metadata_reduced, by="accession")
# fwrite(as.data.frame(grpout), "Results/IPK_All_domesticated_grp_200_6.txt", sep="\t", row.names=T)
# 
# grp_200_7<-find.clusters(genind_obj, max.n.clust=50,  n.pca=200, n.clust=7) # 200 PCs, 7 K
# 
# grpout<-as.data.frame(grp_200_7$grp)
# grpout$accession<-rownames(grpout)
# grpout<-left_join(grpout, metadata_reduced, by="accession")
# fwrite(as.data.frame(grpout), "Results/IPK_All_domesticated_grp_200_7.txt", sep="\t", row.names=T)
#
# grp_200_8<-find.clusters(genind_obj, max.n.clust=50,  n.pca=200, n.clust=8) # 200 PCs, 8 K
# 
# grpout<-as.data.frame(grp_200_8$grp)
# grpout$accession<-rownames(grpout)
# grpout<-left_join(grpout, metadata_reduced, by="accession")
# fwrite(as.data.frame(grpout), "Results/IPK_All_domesticated_grp_200_8.txt", sep="\t", row.names=T)
# 
# grp_200_9<-find.clusters(genind_obj, max.n.clust=50,  n.pca=200, n.clust=9) # 200 PCs, 9 K
# 
# grpout<-as.data.frame(grp_200_9$grp)
# grpout$accession<-rownames(grpout)
# grpout<-left_join(grpout, metadata_reduced, by="accession")
# fwrite(as.data.frame(grpout), "Results/IPK_All_domesticated_grp_200_9.txt", sep="\t", row.names=T)



# Add the cluster assignments for each k to the metadata

###################################### k=2


hmp=read.table("Data/IPK_domesticated_5H_20missing_MAC.hmp.txt",stringsAsFactors=F, na.strings = "N", header = T, sep="\t", comment.char="")

metadata <- fread("Results/IPK_All_domesticated_grp_200_2.txt", na.strings = "")
metadata$V1<-NULL

table(metadata$accession %in% names(hmp)[-c(1:11)])
table(names(hmp)[-c(1:11)] %in% metadata$accession) # 1 false

setdiff(names(hmp)[-c(1:11)], metadata$accession) # remove "BRIDGE_HOR_10349"

hmp$BRIDGE_HOR_10349<-NULL

names(metadata)[1]<-"Hap_cluster"
metadata$Hap_cluster<-as.factor(metadata$Hap_cluster)

# Remove outliers
metadata%<>%filter(accession != "BRIDGE_HOR_7066", accession != "BRIDGE_HOR_21050", accession != "BRIDGE_HOR_16618")

SamplesKeep <- names(hmp)[-c(1:11)]
metadata_reduced <- metadata[metadata$accession%in%SamplesKeep,]

# Read metadata
barn<-fread("Metadata.txt", sep="\t")
barn$Haplotype[barn$Haplotype == "Haplotype 1"] <- "Hap 1"
barn$Haplotype[barn$Haplotype == "Haplotype 2"] <- "Hap 2"
barn$Haplotype[barn$Haplotype == "Haplotype 3"] <- "Hap 3"
barn$Haplotype[barn$Haplotype == "Haplotype 4"] <- "Hap 4"

# Merge datasets
names(barn)[1]<-"BARN_name"
metadata_reduced<-left_join(metadata_reduced, barn, by="BARN_name")

metadata_reduced<-as.data.frame(metadata_reduced)
metadata_reduced[is.na(metadata_reduced$Haplotype),16:21]="NA" # Facilitate plotting by changing potential <NA> to NA

names(hmp)[1]<-"accession"
loc<-transpose_df_Col1(hmp[,-c(2:11)], "accession")

loc<-left_join(metadata_reduced, loc)

locus<-loc[,-c(1:22)]
ind <- names(hmp)[-c(1:11)]

table(names(hmp)[-c(1:11)]  == loc$accession)

# Make genind object
genind_obj <- df2genind(locus, ploidy = 2, ind.names = ind,  sep = "")
genind_obj

X <- scaleGen(genind_obj, NA.method="mean")

# Make PCA
pca1 <- dudi.pca(X,scannf=FALSE, center=T, nf=3)
eig.val <- get_eigenvalue(pca1) # To get the % for the axes

df<-as.data.frame(pca1$li)
df$accession<-rownames(df)
df<-left_join(df, metadata_reduced) # Add passport information


k2=df

fwrite(k2, "PCA_all_domesticated_k2.txt", sep="\t")



#################### k=3
hmp=read.table("Data/IPK_domesticated_5H_20missing_MAC.hmp.txt",stringsAsFactors=F, na.strings = "N", header = T, sep="\t", comment.char="")

metadata <- fread("Results/IPK_All_domesticated_grp_200_3.txt", na.strings = "")
metadata$V1<-NULL

table(metadata$accession %in% names(hmp)[-c(1:11)])
table(names(hmp)[-c(1:11)] %in% metadata$accession) # 1 false

setdiff(names(hmp)[-c(1:11)], metadata$accession) # remove "BRIDGE_HOR_10349"

hmp$BRIDGE_HOR_10349<-NULL

names(metadata)[1]<-"Hap_cluster"
metadata$Hap_cluster<-as.factor(metadata$Hap_cluster)

# Remove outliers
metadata%<>%filter(accession != "BRIDGE_HOR_7066", accession != "BRIDGE_HOR_21050", accession != "BRIDGE_HOR_16618")

SamplesKeep <- names(hmp)[-c(1:11)]
metadata_reduced <- metadata[metadata$accession%in%SamplesKeep,]

# Read metadata
barn<-fread("Metadata.txt", sep="\t")
barn$Haplotype[barn$Haplotype == "Haplotype 1"] <- "Hap 1"
barn$Haplotype[barn$Haplotype == "Haplotype 2"] <- "Hap 2"
barn$Haplotype[barn$Haplotype == "Haplotype 3"] <- "Hap 3"
barn$Haplotype[barn$Haplotype == "Haplotype 4"] <- "Hap 4"

# Merge datasets
names(barn)[1]<-"BARN_name"
metadata_reduced<-left_join(metadata_reduced, barn, by="BARN_name")

metadata_reduced<-as.data.frame(metadata_reduced)
metadata_reduced[is.na(metadata_reduced$Haplotype),16:21]="NA" # Facilitate plotting by changing potential <NA> to NA

names(hmp)[1]<-"accession"
loc<-transpose_df_Col1(hmp[,-c(2:11)], "accession")

loc<-left_join(metadata_reduced, loc)

locus<-loc[,-c(1:22)]
ind <- names(hmp)[-c(1:11)]

table(names(hmp)[-c(1:11)]  == loc$accession)

# Make genind object
genind_obj <- df2genind(locus, ploidy = 2, ind.names = ind,  sep = "")
genind_obj

X <- scaleGen(genind_obj, NA.method="mean")

# Make PCA
pca1 <- dudi.pca(X,scannf=FALSE, center=T, nf=3)
eig.val <- get_eigenvalue(pca1) # To get the % for the axes

df<-as.data.frame(pca1$li)
df$accession<-rownames(df)
df<-left_join(df, metadata_reduced) # Add passport information


k3=df

fwrite(k3, "PCA_all_domesticated_k3.txt", sep="\t")




######################### k=4

hmp=read.table("Data/IPK_domesticated_5H_20missing_MAC.hmp.txt",stringsAsFactors=F, na.strings = "N", header = T, sep="\t", comment.char="")

metadata <- fread("Results/IPK_All_domesticated_grp_200_4.txt", na.strings = "")
metadata$V1<-NULL

table(metadata$accession %in% names(hmp)[-c(1:11)])
table(names(hmp)[-c(1:11)] %in% metadata$accession) # 1 false

setdiff(names(hmp)[-c(1:11)], metadata$accession) # remove "BRIDGE_HOR_10349"

hmp$BRIDGE_HOR_10349<-NULL

names(metadata)[1]<-"Hap_cluster"
metadata$Hap_cluster<-as.factor(metadata$Hap_cluster)

# Remove outliers
metadata%<>%filter(accession != "BRIDGE_HOR_7066", accession != "BRIDGE_HOR_21050", accession != "BRIDGE_HOR_16618")

SamplesKeep <- names(hmp)[-c(1:11)]
metadata_reduced <- metadata[metadata$accession%in%SamplesKeep,]

# Read metadata
barn<-fread("Metadata.txt", sep="\t")
barn$Haplotype[barn$Haplotype == "Haplotype 1"] <- "Hap 1"
barn$Haplotype[barn$Haplotype == "Haplotype 2"] <- "Hap 2"
barn$Haplotype[barn$Haplotype == "Haplotype 3"] <- "Hap 3"
barn$Haplotype[barn$Haplotype == "Haplotype 4"] <- "Hap 4"

# Merge datasets
names(barn)[1]<-"BARN_name"
metadata_reduced<-left_join(metadata_reduced, barn, by="BARN_name")

metadata_reduced<-as.data.frame(metadata_reduced)
metadata_reduced[is.na(metadata_reduced$Haplotype),16:21]="NA" # Facilitate plotting by changing potential <NA> to NA

names(hmp)[1]<-"accession"
loc<-transpose_df_Col1(hmp[,-c(2:11)], "accession")

loc<-left_join(metadata_reduced, loc)

locus<-loc[,-c(1:22)]
ind <- names(hmp)[-c(1:11)]

table(names(hmp)[-c(1:11)]  == loc$accession)

# Make genind object
genind_obj <- df2genind(locus, ploidy = 2, ind.names = ind,  sep = "")
genind_obj

X <- scaleGen(genind_obj, NA.method="mean")

# Make PCA
pca1 <- dudi.pca(X,scannf=FALSE, center=T, nf=3)
eig.val <- get_eigenvalue(pca1) # To get the % for the axes

df<-as.data.frame(pca1$li)
df$accession<-rownames(df)
df<-left_join(df, metadata_reduced) # Add passport information

k4=df

fwrite(k4, "PCA_all_domesticated_k4.txt", sep="\t")




########################### k=5

hmp=read.table("Data/IPK_domesticated_5H_20missing_MAC.hmp.txt",stringsAsFactors=F, na.strings = "N", header = T, sep="\t", comment.char="")

metadata <- fread("Results/IPK_All_domesticated_grp_200_5.txt", na.strings = "")
metadata$V1<-NULL

table(metadata$accession %in% names(hmp)[-c(1:11)])
table(names(hmp)[-c(1:11)] %in% metadata$accession) # 1 false

setdiff(names(hmp)[-c(1:11)], metadata$accession) # remove "BRIDGE_HOR_10349"

hmp$BRIDGE_HOR_10349<-NULL

names(metadata)[1]<-"Hap_cluster"
metadata$Hap_cluster<-as.factor(metadata$Hap_cluster)

# Remove outliers
metadata%<>%filter(accession != "BRIDGE_HOR_7066", accession != "BRIDGE_HOR_21050", accession != "BRIDGE_HOR_16618")

SamplesKeep <- names(hmp)[-c(1:11)]
metadata_reduced <- metadata[metadata$accession%in%SamplesKeep,]

# Read metadata
barn<-fread("Metadata.txt", sep="\t")
barn$Haplotype[barn$Haplotype == "Haplotype 1"] <- "Hap 1"
barn$Haplotype[barn$Haplotype == "Haplotype 2"] <- "Hap 2"
barn$Haplotype[barn$Haplotype == "Haplotype 3"] <- "Hap 3"
barn$Haplotype[barn$Haplotype == "Haplotype 4"] <- "Hap 4"

# Merge datasets
names(barn)[1]<-"BARN_name"
metadata_reduced<-left_join(metadata_reduced, barn, by="BARN_name")

metadata_reduced<-as.data.frame(metadata_reduced)
metadata_reduced[is.na(metadata_reduced$Haplotype),16:21]="NA" # Facilitate plotting by changing potential <NA> to NA

names(hmp)[1]<-"accession"
loc<-transpose_df_Col1(hmp[,-c(2:11)], "accession")

loc<-left_join(metadata_reduced, loc)

locus<-loc[,-c(1:22)]
ind <- names(hmp)[-c(1:11)]

table(names(hmp)[-c(1:11)]  == loc$accession)

# Make genind object
genind_obj <- df2genind(locus, ploidy = 2, ind.names = ind,  sep = "")
genind_obj

X <- scaleGen(genind_obj, NA.method="mean")

# Make PCA
pca1 <- dudi.pca(X,scannf=FALSE, center=T, nf=3)
eig.val <- get_eigenvalue(pca1) # To get the % for the axes

df<-as.data.frame(pca1$li)
df$accession<-rownames(df)
df<-left_join(df, metadata_reduced) # Add passport information

k5=df

fwrite(k5, "PCA_all_domesticated_k2.txt", sep="\t")



################################### k=5

hmp=read.table("Data/IPK_domesticated_5H_20missing_MAC.hmp.txt",stringsAsFactors=F, na.strings = "N", header = T, sep="\t", comment.char="")

metadata <- fread("Results/IPK_All_domesticated_grp_200_6.txt", na.strings = "")
metadata$V1<-NULL

table(metadata$accession %in% names(hmp)[-c(1:11)])
table(names(hmp)[-c(1:11)] %in% metadata$accession) # 1 false

setdiff(names(hmp)[-c(1:11)], metadata$accession) # remove "BRIDGE_HOR_10349"

hmp$BRIDGE_HOR_10349<-NULL

names(metadata)[1]<-"Hap_cluster"
metadata$Hap_cluster<-as.factor(metadata$Hap_cluster)

# Remove outliers
metadata%<>%filter(accession != "BRIDGE_HOR_7066", accession != "BRIDGE_HOR_21050", accession != "BRIDGE_HOR_16618")

SamplesKeep <- names(hmp)[-c(1:11)]
metadata_reduced <- metadata[metadata$accession%in%SamplesKeep,]

# Read metadata
barn<-fread("Metadata.txt", sep="\t")
barn$Haplotype[barn$Haplotype == "Haplotype 1"] <- "Hap 1"
barn$Haplotype[barn$Haplotype == "Haplotype 2"] <- "Hap 2"
barn$Haplotype[barn$Haplotype == "Haplotype 3"] <- "Hap 3"
barn$Haplotype[barn$Haplotype == "Haplotype 4"] <- "Hap 4"

# Merge datasets
names(barn)[1]<-"BARN_name"
metadata_reduced<-left_join(metadata_reduced, barn, by="BARN_name")

metadata_reduced<-as.data.frame(metadata_reduced)
metadata_reduced[is.na(metadata_reduced$Haplotype),16:21]="NA" # Facilitate plotting by changing potential <NA> to NA

names(hmp)[1]<-"accession"
loc<-transpose_df_Col1(hmp[,-c(2:11)], "accession")

loc<-left_join(metadata_reduced, loc)

locus<-loc[,-c(1:22)]
ind <- names(hmp)[-c(1:11)]

table(names(hmp)[-c(1:11)]  == loc$accession)

# Make genind object
genind_obj <- df2genind(locus, ploidy = 2, ind.names = ind,  sep = "")
genind_obj

X <- scaleGen(genind_obj, NA.method="mean")

# Make PCA
pca1 <- dudi.pca(X,scannf=FALSE, center=T, nf=3)
eig.val <- get_eigenvalue(pca1) # To get the % for the axes

df<-as.data.frame(pca1$li)
df$accession<-rownames(df)
df<-left_join(df, metadata_reduced) # Add passport information

k6=df

fwrite(k6, "PCA_all_domesticated_k6.txt", sep="\t")




################################ k=7

hmp=read.table("Data/IPK_domesticated_5H_20missing_MAC.hmp.txt",stringsAsFactors=F, na.strings = "N", header = T, sep="\t", comment.char="")

metadata <- fread("Results/IPK_All_domesticated_grp_200_7.txt", na.strings = "")
metadata$V1<-NULL

table(metadata$accession %in% names(hmp)[-c(1:11)])
table(names(hmp)[-c(1:11)] %in% metadata$accession) # 1 false

setdiff(names(hmp)[-c(1:11)], metadata$accession) # remove "BRIDGE_HOR_10349"

hmp$BRIDGE_HOR_10349<-NULL

names(metadata)[1]<-"Hap_cluster"
metadata$Hap_cluster<-as.factor(metadata$Hap_cluster)

# Remove outliers
metadata%<>%filter(accession != "BRIDGE_HOR_7066", accession != "BRIDGE_HOR_21050", accession != "BRIDGE_HOR_16618")

SamplesKeep <- names(hmp)[-c(1:11)]
metadata_reduced <- metadata[metadata$accession%in%SamplesKeep,]

# Read metadata
barn<-fread("Metadata.txt", sep="\t")
barn$Haplotype[barn$Haplotype == "Haplotype 1"] <- "Hap 1"
barn$Haplotype[barn$Haplotype == "Haplotype 2"] <- "Hap 2"
barn$Haplotype[barn$Haplotype == "Haplotype 3"] <- "Hap 3"
barn$Haplotype[barn$Haplotype == "Haplotype 4"] <- "Hap 4"

# Merge datasets
names(barn)[1]<-"BARN_name"
metadata_reduced<-left_join(metadata_reduced, barn, by="BARN_name")

metadata_reduced<-as.data.frame(metadata_reduced)
metadata_reduced[is.na(metadata_reduced$Haplotype),16:21]="NA" # Facilitate plotting by changing potential <NA> to NA

names(hmp)[1]<-"accession"
loc<-transpose_df_Col1(hmp[,-c(2:11)], "accession")

loc<-left_join(metadata_reduced, loc)

locus<-loc[,-c(1:22)]
ind <- names(hmp)[-c(1:11)]

table(names(hmp)[-c(1:11)]  == loc$accession)

# Make genind object
genind_obj <- df2genind(locus, ploidy = 2, ind.names = ind,  sep = "")
genind_obj

X <- scaleGen(genind_obj, NA.method="mean")

# Make PCA
pca1 <- dudi.pca(X,scannf=FALSE, center=T, nf=3)
eig.val <- get_eigenvalue(pca1) # To get the % for the axes

df<-as.data.frame(pca1$li)
df$accession<-rownames(df)
df<-left_join(df, metadata_reduced) # Add passport information

k7=df

fwrite(k7, "PCA_all_domesticated_k7.txt", sep="\t")



################################ k=8

hmp=read.table("Data/IPK_domesticated_5H_20missing_MAC.hmp.txt",stringsAsFactors=F, na.strings = "N", header = T, sep="\t", comment.char="")

metadata <- fread("Results/IPK_All_domesticated_grp_200_8.txt", na.strings = "")
metadata$V1<-NULL

table(metadata$accession %in% names(hmp)[-c(1:11)])
table(names(hmp)[-c(1:11)] %in% metadata$accession) # 1 false

setdiff(names(hmp)[-c(1:11)], metadata$accession) # remove "BRIDGE_HOR_10349"

hmp$BRIDGE_HOR_10349<-NULL

names(metadata)[1]<-"Hap_cluster"
metadata$Hap_cluster<-as.factor(metadata$Hap_cluster)

# Remove outliers
metadata%<>%filter(accession != "BRIDGE_HOR_7066", accession != "BRIDGE_HOR_21050", accession != "BRIDGE_HOR_16618")

SamplesKeep <- names(hmp)[-c(1:11)]
metadata_reduced <- metadata[metadata$accession%in%SamplesKeep,]

# Read metadata
barn<-fread("Metadata.txt", sep="\t")
barn$Haplotype[barn$Haplotype == "Haplotype 1"] <- "Hap 1"
barn$Haplotype[barn$Haplotype == "Haplotype 2"] <- "Hap 2"
barn$Haplotype[barn$Haplotype == "Haplotype 3"] <- "Hap 3"
barn$Haplotype[barn$Haplotype == "Haplotype 4"] <- "Hap 4"

# Merge datasets
names(barn)[1]<-"BARN_name"
metadata_reduced<-left_join(metadata_reduced, barn, by="BARN_name")

metadata_reduced<-as.data.frame(metadata_reduced)
metadata_reduced[is.na(metadata_reduced$Haplotype),16:21]="NA" # Facilitate plotting by changing potential <NA> to NA

names(hmp)[1]<-"accession"
loc<-transpose_df_Col1(hmp[,-c(2:11)], "accession")

loc<-left_join(metadata_reduced, loc)

locus<-loc[,-c(1:22)]
ind <- names(hmp)[-c(1:11)]

table(names(hmp)[-c(1:11)]  == loc$accession)

# Make genind object
genind_obj <- df2genind(locus, ploidy = 2, ind.names = ind,  sep = "")
genind_obj

X <- scaleGen(genind_obj, NA.method="mean")

# Make PCA
pca1 <- dudi.pca(X,scannf=FALSE, center=T, nf=3)
eig.val <- get_eigenvalue(pca1) # To get the % for the axes

df<-as.data.frame(pca1$li)
df$accession<-rownames(df)
df<-left_join(df, metadata_reduced) # Add passport information

k8=df

fwrite(k8, "PCA_all_domesticated_k8.txt", sep="\t")




################################ k=9

hmp=read.table("Data/IPK_domesticated_5H_20missing_MAC.hmp.txt",stringsAsFactors=F, na.strings = "N", header = T, sep="\t", comment.char="")

metadata <- fread("Results/IPK_All_domesticated_grp_200_9.txt", na.strings = "")
metadata$V1<-NULL

table(metadata$accession %in% names(hmp)[-c(1:11)])
table(names(hmp)[-c(1:11)] %in% metadata$accession) # 1 false

setdiff(names(hmp)[-c(1:11)], metadata$accession) # remove "BRIDGE_HOR_10349"

hmp$BRIDGE_HOR_10349<-NULL

names(metadata)[1]<-"Hap_cluster"
metadata$Hap_cluster<-as.factor(metadata$Hap_cluster)

# Remove outliers
metadata%<>%filter(accession != "BRIDGE_HOR_7066", accession != "BRIDGE_HOR_21050", accession != "BRIDGE_HOR_16618")

SamplesKeep <- names(hmp)[-c(1:11)]
metadata_reduced <- metadata[metadata$accession%in%SamplesKeep,]

# Read metadata
barn<-fread("Metadata.txt", sep="\t")
barn$Haplotype[barn$Haplotype == "Haplotype 1"] <- "Hap 1"
barn$Haplotype[barn$Haplotype == "Haplotype 2"] <- "Hap 2"
barn$Haplotype[barn$Haplotype == "Haplotype 3"] <- "Hap 3"
barn$Haplotype[barn$Haplotype == "Haplotype 4"] <- "Hap 4"

# Merge datasets
names(barn)[1]<-"BARN_name"
metadata_reduced<-left_join(metadata_reduced, barn, by="BARN_name")

metadata_reduced<-as.data.frame(metadata_reduced)
metadata_reduced[is.na(metadata_reduced$Haplotype),16:21]="NA" # Facilitate plotting by changing potential <NA> to NA

names(hmp)[1]<-"accession"
loc<-transpose_df_Col1(hmp[,-c(2:11)], "accession")

loc<-left_join(metadata_reduced, loc)

locus<-loc[,-c(1:22)]
ind <- names(hmp)[-c(1:11)]

table(names(hmp)[-c(1:11)]  == loc$accession)

# Make genind object
genind_obj <- df2genind(locus, ploidy = 2, ind.names = ind,  sep = "")
genind_obj

X <- scaleGen(genind_obj, NA.method="mean")

# Make PCA
pca1 <- dudi.pca(X,scannf=FALSE, center=T, nf=3)
eig.val <- get_eigenvalue(pca1) # To get the % for the axes

df<-as.data.frame(pca1$li)
df$accession<-rownames(df)
df<-left_join(df, metadata_reduced) # Add passport information

k9=df

fwrite(k9, "PCA_all_domesticated_k9.txt", sep="\t")









#######################################################
## Make a figure with cluster memberships for each k ##
#######################################################


library(ggforce)

k2<-fread("PCA_all_domesticated_k2.txt")
k3<-fread("PCA_all_domesticated_k3.txt")
k4<-fread("PCA_all_domesticated_k4.txt")
k5<-fread("PCA_all_domesticated_k5.txt")
k6<-fread("PCA_all_domesticated_k6.txt")
k7<-fread("PCA_all_domesticated_k7.txt")
k8<-fread("PCA_all_domesticated_k8.txt")
k9<-fread("PCA_all_domesticated_k9.txt")

k2$Hap_cluster<-as.factor(k2$Hap_cluster)
k3$Hap_cluster<-as.factor(k3$Hap_cluster)
k4$Hap_cluster<-as.factor(k4$Hap_cluster)
k5$Hap_cluster<-as.factor(k5$Hap_cluster)
k6$Hap_cluster<-as.factor(k6$Hap_cluster)
k7$Hap_cluster<-as.factor(k7$Hap_cluster)
k8$Hap_cluster<-as.factor(k8$Hap_cluster)
k9$Hap_cluster<-as.factor(k9$Hap_cluster)



f2<-ggplot(k2 , aes(x=Axis1, y=Axis2, color=Hap_cluster) )+ geom_point(alpha=0.3, size=0.5) +
  scale_color_manual(values = colors)+
  xlab(paste0("PC1 (", round(eig.val$variance.percent[1], 2), "%)"))+ ylab(paste0("PC2 (", round(eig.val$variance.percent[2], 2), "%)"))+
  theme_minimal() + theme(legend.position = "none") + ggtitle("k=2")+coord_fixed() +  theme(axis.text=element_text(size=6),axis.title=element_text(size=7), plot.title = element_text(size = 8))

f3<-ggplot(k3 , aes(x=Axis1, y=Axis2, color=Hap_cluster) )+ geom_point(alpha=0.3, size=0.5) +
  scale_color_manual(values = colors)+
  xlab(paste0("PC1 (", round(eig.val$variance.percent[1], 2), "%)"))+ ylab(paste0("PC2 (", round(eig.val$variance.percent[2], 2), "%)"))+
  theme_minimal() + theme(legend.position = "none") + ggtitle("k=3") +coord_fixed()+  theme(axis.text=element_text(size=6),axis.title=element_text(size=7), plot.title = element_text(size = 8))

f4<-ggplot(k4 , aes(x=Axis1, y=Axis2, color=Hap_cluster) )+ geom_point(alpha=0.3, size=0.5) +
  scale_color_manual(values = colors)+
  xlab(paste0("PC1 (", round(eig.val$variance.percent[1], 2), "%)"))+ ylab(paste0("PC2 (", round(eig.val$variance.percent[2], 2), "%)"))+
  theme_minimal() + theme(legend.position = "none") + ggtitle("k=4") +coord_fixed()+  theme(axis.text=element_text(size=6),axis.title=element_text(size=7), plot.title = element_text(size = 8))

f5<-ggplot(k5 , aes(x=Axis1, y=Axis2, color=Hap_cluster) )+ geom_point(alpha=0.3, size=0.5) +
  scale_color_manual(values = colors)+
  xlab(paste0("PC1 (", round(eig.val$variance.percent[1], 2), "%)"))+ ylab(paste0("PC2 (", round(eig.val$variance.percent[2], 2), "%)"))+
  theme_minimal() + theme(legend.position = "none") + ggtitle("k=5") +coord_fixed()+  theme(axis.text=element_text(size=6),axis.title=element_text(size=7), plot.title = element_text(size = 8))

f6<-ggplot(k6 , aes(x=Axis1, y=Axis2, color=Hap_cluster) )+ geom_point(alpha=0.3, size=0.5) +
  scale_color_manual(values = colors)+
  xlab(paste0("PC1 (", round(eig.val$variance.percent[1], 2), "%)"))+ ylab(paste0("PC2 (", round(eig.val$variance.percent[2], 2), "%)"))+
  theme_minimal() + theme(legend.position = "none") + ggtitle("k=6") +coord_fixed()+  theme(axis.text=element_text(size=6),axis.title=element_text(size=7), plot.title = element_text(size = 8))

f7<-ggplot(k7 , aes(x=Axis1, y=Axis2, color=Hap_cluster) )+ geom_point(alpha=0.3, size=0.5) +
  scale_color_manual(values = colors)+
  xlab(paste0("PC1 (", round(eig.val$variance.percent[1], 2), "%)"))+ ylab(paste0("PC2 (", round(eig.val$variance.percent[2], 2), "%)"))+
  theme_minimal() + theme(legend.position = "none") + ggtitle("k=7")+
  geom_ellipse(aes(x0=-2.5, y0=-27, a=8, b=15, angle=0.2, m1=2), color="black") +coord_fixed() +  theme(axis.text=element_text(size=6),axis.title=element_text(size=7), plot.title = element_text(size = 8))

f8<-ggplot(k8 , aes(x=Axis1, y=Axis2, color=Hap_cluster) )+ geom_point(alpha=0.3, size=0.5) +
  scale_color_manual(values = colors)+
  xlab(paste0("PC1 (", round(eig.val$variance.percent[1], 2), "%)"))+ ylab(paste0("PC2 (", round(eig.val$variance.percent[2], 2), "%)"))+
  theme_minimal() + theme(legend.position = "none") + ggtitle("k=8")+
  geom_ellipse(aes(x0=-2.5, y0=-27, a=8, b=15, angle=0.2, m1=2), color="black") +coord_fixed()  +  theme(axis.text=element_text(size=6),axis.title=element_text(size=7), plot.title = element_text(size = 8))

f9<-ggplot(k9 , aes(x=Axis1, y=Axis2, color=Hap_cluster) )+ geom_point(alpha=0.3, size=0.5) +
  scale_color_manual(values = colors)+
  xlab(paste0("PC1 (", round(eig.val$variance.percent[1], 2), "%)"))+ ylab(paste0("PC2 (", round(eig.val$variance.percent[2], 2), "%)"))+
  theme_minimal() + theme(legend.position = "none") + ggtitle("k=9")+
  geom_ellipse(aes(x0=-2.5, y0=-27, a=8, b=15, angle=0.2, m1=2), color="black") +coord_fixed() +  theme(axis.text=element_text(size=6),axis.title=element_text(size=7), plot.title = element_text(size = 8))



tiff("ESM_19.tiff", height=8, width=17.4, res=600, units="cm")
figure<-ggarrange(f2, f3, f4, f5, f6, f7, f8, f9, nrow=2, ncol=4)

figure


dev.off()


###################################################################################################################################
### Make a table including all information and the cluster assignment to save for e.g. making the maps of haplotype distribution ##
###################################################################################################################################

df<-as.data.table(df)


# Assign haplotypes to clusters based on known haplotypes in the cluster. You will need to adjust the assignments because cluster numbers will be different after you've run the k-means clustering

df$Hap_cluster<-as.character(df$Hap_cluster)

df$Hap_cluster[df$Hap_cluster == "1"] <- "Other 1"
df$Hap_cluster[df$Hap_cluster == "2"] <- "Other 2"
df$Hap_cluster[df$Hap_cluster == "3"] <- "Hap 2"
df$Hap_cluster[df$Hap_cluster == "4"] <- "Hap 3"
df$Hap_cluster[df$Hap_cluster == "5"] <- "Hap 4"
df$Hap_cluster[df$Hap_cluster == "6"] <- "Hap 1"
df$Hap_cluster[df$Hap_cluster == "7"] <- "Other 3"


fwrite(df, "PCA_all_domesticated_clusters.txt", sep="\t", row.names=T)

