#####################################################
## PCAs of the IPK genebank barley collection data ##
#####################################################

# Note: To analyze the IPK genebank data, we utilized WGS data generated for all accessions. the original dataset was published in Milner et al., 2019. Here we used an updated dataset with all variants mapped to Morex V3 (Mascher et al., 2021) provided by Martin Mascher, IPK Gatersleben. 

source("https://raw.githubusercontent.com/rwonneberger/R_functions/main/Ronja_functions.R")
library(adegenet)
library(factoextra)


# The 5H region of interest is located at ~70 - 320 Mbp in Barke. We extracted 70000000-70005000 and 320000000-320005000 from Barke and blasted the sequences against Morex V3 (Mascher et al., 2021).
# The start is ca 66.2 Mbp in Morex, we selected 67 Mbp as a more conservative approach. The end is ca 312.2 Mbp in Morex, we selected 312 Mbp as a more conservative approach.
# We only considered accessions that were "domesticated" according to their passport information.
# We filtered the dataset to remove indels, and only kept SNPs with a minor allele count of > 1 and missing data < 20%.
# The final dataset contained 6536 SNPs and 19838 individuals.
# This script requires the described dataset as input, which will need to be generated by the Filtering_genebank_SNP_data.sh script. In addition, the user will need the passport data of the gene bank collection which can be found in Miler et al., 2019.


hmp=read.table("Data/IPK_domesticated_5H_20missing_MAC.hmp.txt",stringsAsFactors=F, na.strings = "N", header = T, sep="\t", comment.char="")
 
metadata <- fread("IPK_domesticated.txt") # To be generated by user. See Milner et al., 2019


# Read in a metadata file generated from Online Resource 1
barn<-fread("Metadata.txt", sep="\t")
names(barn)[1]<-"BARN_name"

# Merge genebank and BARN panel passport data
metadata_reduced<-left_join(metadata, barn, by="BARN_name")

metadata_reduced<-as.data.frame(metadata_reduced)

names(hmp)[1]<-"accession"
loc<-transpose_df_Col1(hmp[,-c(2:11)], "accession")

loc<-left_join(metadata_reduced, loc)

locus<-loc[,-c(1:21)]
ind <- names(hmp)[-c(1:11)] 

table(names(hmp)[-c(1:11)]  == loc$accession)

# Make a genind project
genind_obj <- df2genind(locus, ploidy = 2, ind.names = ind,  sep = "")
genind_obj

X <- scaleGen(genind_obj, NA.method="mean")

# Run PCA
pca1 <- dudi.pca(X,scannf=FALSE, center=T, nf=3)
eig.val <- get_eigenvalue(pca1) # To get the % for the axes


df<-as.data.frame(pca1$li)
df$accession<-rownames(df)
df<-left_join(df, metadata_reduced)


# Prepare data for plotting, deal with missing data
emptycells<-which(is.na(df$Haplotype))
df$Haplotype[emptycells ]<-"unknown"
str(df$Haplotype)
df$Haplotype<-as.factor(df$Haplotype)
factororder<-sort(table(df$Haplotype), decreasing=T)
df$Haplotype<-factor(df$Haplotype, levels = names(factororder))
df%<>%arrange(Haplotype) 

index_for_grey<-which(names(factororder) == "unknown")

colors=c('#E69F00', '#009E73', '#56B4E9', "#D55E00" )

colors <- append(colors, 'lightgrey', after=index_for_grey-1)


p1<-ggplot(df%>%arrange((Haplotype)) , aes(x=Axis1, y=Axis2, color=Haplotype) )+ geom_point(alpha=0.3, size=0.5) +
  scale_color_manual(values = colors)+ 
  xlab(paste0("PC1 (", round(eig.val$variance.percent[1], 2), "%)"))+ ylab(paste0("PC2 (", round(eig.val$variance.percent[2], 2), "%)"))+
  theme_minimal() +
  theme(axis.text=element_text(size=6),
        axis.title=element_text(size=7))+ 
  theme(legend.title = element_text(size=7),
        legend.text = element_text(size=6))+ 
  guides(colour = guide_legend(override.aes = list(alpha=1)))+ 
  labs(color = "Chr 5H haplotype")


tiff("Fig9.tiff", height=6 ,width=8.4, units="cm", res=600)
p1

dev.off()





## Color one geographic origin at a time
regions<-unique(df$Region2)
regions<-regions[!regions == "unknown"]

p=list()
count = 0

for (i in regions){
  count = count +1
  
emptycells<-which(is.na(df$Region2))
df$Region2[emptycells ]<-"unknown"

df$coloring<-"none"
df[df$Region2 == i,]$coloring<-"yes"

colors=c("grey", "#D55E00")


p[[i]]<-ggplot(df%>%arrange(coloring) , aes(x=Axis1, y=Axis2, color=coloring) )+ geom_point(alpha=0.3, size=0.5) +
  scale_color_manual(values = colors)+ 
  xlab(paste0("PC1 (", round(eig.val$variance.percent[1], 2), "%)"))+ ylab(paste0("PC2 (", round(eig.val$variance.percent[2], 2), "%)"))+
  theme_minimal() +
  theme(axis.text=element_text(size=6),
        axis.title=element_text(size=7))+ 
  theme(legend.title = element_text(size=7),
        legend.text = element_text(size=6))+ 
  guides(colour = guide_legend(override.aes = list(alpha=1)))+ 
  theme(legend.position = "none")+
  ggtitle(i)

}

tiff("ESM_22.tiff", height=30 ,width=17.4, units="cm", res=600)
ggarrange( p[[1]], p[[2]], p[[3]], p[[4]], p[[5]], p[[6]], p[[7]],  ncol=2, nrow=4, labels = c("a", "b", "c", "d", "e", "f", "g"))

dev.off()



